### PR TITLE
add enrollment status metric for cirrus

### PR DESCRIPTION
**Pull request by jeddai**
### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [x] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/nimbus/src/tests/helpers.rs [hunk 1]
```diff
@@ -49,6 +48,50 @@ impl Default for NimbusTargetingHelper {
...
+        let key = "enrollment_status";
+        let mut state = self.state.lock().unwrap();
+        let new = serde_json::to_value(enrollment_status_extras).unwrap();
+        match state.get(key) {
+            Some(v) => {
+                let new_value = v
+                    .as_array()
+                    .unwrap()
+                    .iter()
+                    .chain(new.as_array().unwrap())
```
Comment (brennie) [inline components/nimbus/src/tests/helpers.rs:83]: Can we use as_array_mut() here and modify in-place?

File: components/nimbus/src/tests/helpers.rs [hunk 2]
```diff
@@ -49,6 +48,50 @@ impl Default for NimbusTargetingHelper {
...
+                    .cloned()
+                    .collect::<Vec<serde_json::Value>>();
+                state.insert(key.into(), new_value.into());
+            }
+            None => {
+                state.insert(key.into(), new);
+            }
+        };
```
Comment (brennie) [inline components/nimbus/src/tests/helpers.rs:91]: Prefer the entry API to do only a single lookup.

```rust
map.entry(key)
   .and_modify(|v| /* ... */)
   .or_insert_with(|| /* ... */)
```

File: megazords/cirrus/tests/python-tests/conftest.py [hunk 3]
```diff
@@ -3,10 +3,20 @@
...
-from cirrus import CirrusClient
+from cirrus import CirrusClient, MetricsHandler, EnrollmentStatusExtraDef
 from fml import FmlClient
 
 
+class TestMetricsHandler(MetricsHandler):
+    recordings = []
+
+    @classmethod
+    def record_enrollment_statuses(self, enrollment_status_extras: [EnrollmentStatusExtraDef]):
```
Comment (brennie) [inline megazords/cirrus/tests/python-tests/conftest.py:14]: The correct annotation is `list[T]`
```suggestion
    def record_enrollment_statuses(self, enrollment_status_extras: list[EnrollmentStatusExtraDef]):
```

File: megazords/cirrus/tests/python-tests/conftest.py [hunk 4]
```diff
@@ -3,10 +3,20 @@
...
+        self.recordings.clear()
+        for enrollment_status_extra in enrollment_status_extras:
+            self.recordings.append(enrollment_status_extra)
```
Comment (brennie) [inline megazords/cirrus/tests/python-tests/conftest.py:17]:
```suggestion
        self.recordings.extend(enrollment_status_extra)
```

File: components/nimbus/src/metrics.rs [hunk 5]
```diff
@@ -40,6 +42,11 @@ impl EnrollmentStatusExtraDef {
     pub fn status(&self) -> &str {
         self.status.as_ref().unwrap()
     }
+
+    #[cfg(not(feature = "stateful"))]
+    pub fn user_id(&self) -> &str {
+        self.user_id.as_ref().unwrap()
+    }
```
[the remark this answers is not in the document]
Comment (jhugman) [reply components/nimbus/src/metrics.rs:49]: I think given that this is test code, that crashing hard is the right thing to do if it's not intended.

File: components/nimbus/src/stateful/nimbus_client.rs [hunk 6]
```diff
@@ -81,7 +81,7 @@ impl NimbusClient {
         db_path: P,
         config: Option<RemoteSettingsConfig>,
         available_randomization_units: AvailableRandomizationUnits,
-        host_metrics: Box<dyn MetricsHandler>,
+        metrics_handler: Box<dyn MetricsHandler>,
```
Comment (jhugman) [inline components/nimbus/src/stateful/nimbus_client.rs:84]: 😻

File: components/nimbus/src/tests/helpers.rs [hunk 7]
```diff
@@ -49,6 +48,50 @@ impl Default for NimbusTargetingHelper {
...
+        }
+    }
+
+    pub fn assert_get_vec_value(&self, key: &str) -> serde_json::Value {
+        self.state.lock().unwrap().get(key).unwrap().clone()
+    }
+}
+
+impl MetricsHandler for TestMetrics {
+    fn record_enrollment_statuses(&self, enrollment_status_extras: Vec<EnrollmentStatusExtraDef>) {
```
Comment (jhugman) [inline components/nimbus/src/tests/helpers.rs:73]: Nit: Would you mind writing a comment on what this code is meant to do?

File: components/nimbus/src/stateless/cirrus_client.rs [hunk 8]
```diff
@@ -76,20 +77,25 @@ pub struct CirrusMutableState {
     experiments: Vec<Experiment>,
 }
 
-#[derive(Default)]
 pub struct CirrusClient {
     app_context: AppContext,
     coenrolling_feature_ids: Vec<String>,
     state: Mutex<CirrusMutableState>,
+    metrics_handler: Arc<Box<dyn MetricsHandler>>,
```
Comment (jhugman) [inline components/nimbus/src/stateless/cirrus_client.rs:84]: Hmm. This is going to be a breaking change for cirrus over in mozilla/experimenter. I wonder if we should migrate that python (or a portion of it) and its tests over here.

**Review by jhugman — APPROVED** [review]
This is looking just fine. I'm approving this, though you should address the nits that I've raised. 

@brennie have you any reason for this not to be r+wc?

File: components/nimbus/src/metrics.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (brennie) [reply components/nimbus/src/metrics.rs:49]: Oops I missed that this part was just for tests. Nevermind then!

**Review by brennie — APPROVED** [review]
I meant to approve this last time but forgot to tick the box. 😅😅

File: components/nimbus/src/tests/helpers.rs (hunk 2 above)
Comment (jeddai) [reply components/nimbus/src/tests/helpers.rs:91]: Oooooh this API is news to me — this is super cool!

File: components/nimbus/src/stateless/cirrus_client.rs (hunk 8 above)
Comment (jeddai) [reply components/nimbus/src/stateless/cirrus_client.rs:84]: That's very possible — we could likely do something similar to what we have set up with the Kotlin/Swift

**Review by eliserichards — APPROVED** [review]
🚀

**Pull request merged**